libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -733,28 +736,59 @@ bool ContractCompiler::visit(ForStatement const& _forStatement)
 		_forStatement.loopExpression()->accept(*this);
 
 	m_context.appendJumpTo(loopStart);
+
+	endVisitLoop(&_forStatement);
+
+	shared_ptr<eth::AssemblyItem> afterFreeing;
+	if (m_breakTagsPops.size() > 0)
+	{
+		afterFreeing = make_shared<eth::AssemblyItem>(m_context.newTag());
+		freeLocalLoopVariables(loopStart, afterFreeing);
+	}
+
 	m_context << loopEnd;
 
-	m_continueTags.pop_back();
-	m_breakTags.pop_back();
+	// For the case where no break is executed:
+	// loop initialization variables have to be freed
+	popBlockScopedVariables(&_forStatement);
+
+	// Location where the break tags jump to
+	if (afterFreeing)
+		m_context << *afterFreeing;
 
 	checker.check();
 	return false;
 }
 
 bool ContractCompiler::visit(Continue const& _continueStatement)
 {
-	CompilerContext::LocationSetter locationSetter(m_context, _continueStatement);
-	if (!m_continueTags.empty())
-		m_context.appendJumpTo(m_continueTags.back());
-	return false;
+	return visitBreakContinue(&_continueStatement);
 }
 
 bool ContractCompiler::visit(Break const& _breakStatement)
 {
-	CompilerContext::LocationSetter locationSetter(m_context, _breakStatement);
-	if (!m_breakTags.empty())
-		m_context.appendJumpTo(m_breakTags.back());
+	return visitBreakContinue(&_breakStatement);
+}
+
+bool ContractCompiler::visitBreakContinue(Statement const* _statement)
+{
+	CompilerContext::LocationSetter locationSetter(m_context, *_statement);
+	solAssert(!m_loops.empty(), "");
+	unsigned sizeOnStack = 0;
+	for (auto _decl : m_loopScopedVariables[m_loops.back()])
+		sizeOnStack += _decl->type()->sizeOnStack();
+	m_breakTagsPops.push_back(std::make_pair<unsigned, eth::AssemblyItem>(
+		unsigned(sizeOnStack),
+		m_context.newTag()
+	));
+	if (dynamic_cast<Break const*>(_statement))
+		m_context << 0;
```

Comment:
Whit is the purpose of these constants?

Reply:
Allowing breaks and continues to share the pop structure. In the end of the pops, a 0 would lead to a jump to outside the loop and a 1 would jump back to the loop condition test.